customers/services/profile.py

```diff
@@ -0,0 +1,128 @@
+                edges {
+                    node {
+                        id
+                        first_name: firstName
+                        last_name: lastName
+                        primary_email: primaryEmail {
+                            email
+                        }
+                    }
+                }
+            }
+        }
+    """
+
+    def __init__(self, request, profile_list_threshold=None, **kwargs):
+        if "config" in kwargs:
+            self.config = kwargs.get("config")
+
+        self.profile_token = self._get_city_profile_token(request)
+        self.profile_list_threshold = (
+            profile_list_threshold or self.profile_list_threshold
+        )
+        self.api_url = self.config.get(PROFILE_API_URL)
+
+    @staticmethod
+    def get_config_template():
+        return {
+            PROFILE_API_URL: str,
+        }
+
+    def _get_city_profile_token(self, request: HttpRequest, nullable: bool = False):
+        api_tokens = request.META.get("HTTP_API_TOKENS")
+        if not api_tokens:
+            raise Exception("Missing tokens")
+
+        api_tokens = json.loads(api_tokens)
+        profile_token = api_tokens.get("https://api.hel.fi/auth/helsinkiprofile")
+        if not nullable and not profile_token:
+            raise Exception("No token found in the request")
+
+        return profile_token
+
+    def get_all_profiles(self) -> Dict[str, CustomerProfile]:
+        query = self.ALL_PROFILES_QUERY
+
+        response = self.query(query)
+        edges = response.get("profiles", {}).get("edges", [])
+
+        # Parse the users received from the Profile Service
+        users = {}
+        for edge in edges:
+            user = self.parse_user_edge(edge)
+            users[user.id] = user
+
+        # Query all the profiles associated to them.
+        # After the loop to do it on a single query.
+        customer_profiles = CustomerProfile.objects.filter(id__in=users.keys())
+
+        # Associate the customer profiles with the fetched from the Profile Service
+        for profile in customer_profiles:
+            user = users[profile.id]
+            profile.user = user
+            users[profile.id] = profile
+
+        return users
+
+    def get_profile(self, id: UUID) -> CustomerProfile:
+        from ..schema import ProfileNode
+
+        global_id = to_global_id(ProfileNode, id)
+
+        query = f"""
+            query GetProfile {{
+                profile(serviceType: BERTH, id: "{global_id}") {{
+                    id
+                    first_name: firstName
+                    last_name: lastName
+                    primary_email: primaryEmail {{
+                        email
+                    }}
+                }}
+            }}
+        """
+
+        response = self.query(query)
+        user = self.parse_user(response.pop("profile"))
+        profile = CustomerProfile.objects.get(id=id)
+        profile.user = user
```

Comment:
AFAIK `profile.user` is the "Django user" why we are overriding that with the object from Hki profile?

Reply:
My reasoning for this was to have a common concise User model witht the fields that we need rather than just dicts. By setting the user to the `CustomerProfile`, we would have access to all the customer-related stuff when needed, like leases, orders, etc... and also the personal information through the `user` object.

Reply:
I think it's dangerous to override a field with something else that it's supposed to be. Could we create a new class called `HkiProfileUser` that holds only the fields form hki profiili (`id, first name, last name, email`). Then this service would expose only those kind of objects. What do you think? Would also like to hear opinion(s) from @ahie and @haavikko ?